storage/command_queue.go

```diff
@@ -87,17 +95,71 @@ func (cq *CommandQueue) GetWait(readOnly bool, wg *sync.WaitGroup, spans ...roac
 			end = start.Next()
 			start = end[:len(start)]
 		}
-		for _, c := range cq.cache.GetOverlaps(start, end) {
-			c := c.Value.(*cmd)
-			// Only add to the wait group if one of the commands isn't read-only.
-			if !readOnly || !c.readOnly {
+		overlaps := cq.cache.GetOverlaps(start, end)
+		if readOnly {
+			// If both commands are read-only, there are no dependencies between them,
+			// so these can be filtered out of the overlapping commands.
+			overlaps = filterReadWrite(overlaps)
+		}
+
+		// Easy cases that can avoid sorting and allocation.
+		switch len(overlaps) {
+		case 0:
+			continue
+		case 1:
+			c := overlaps[0].Value.(*cmd)
+			c.pending = append(c.pending, wg)
+			wg.Add(1)
+			continue
+		}
+
+		// Sort overlapping commands by timestamp and iterate from latest to earliest,
+		// adding the commands' ranges to the RangeGroup to determine gating keyspace
+		// command dependencies. Because all commands are given WaitGroup dependencies
+		// to the most recent commands that they are dependent on, and because of the
+		// causality provided by the Hybrid Logical Clock used to timestamp commands,
```

Comment:
Raft can reorder commands (i.e. there's no guarantee that a command with a higher timestamp executes after another pending command with a lower timestamp). Won't that cause issues?